test/system/src/main/java/io/pravega/test/system/framework/kubernetes/K8sClient.java

```diff
@@ -666,31 +666,63 @@ public void deletePVC(String namespace, String name) {
+    }
 
-        AtomicBoolean shouldRetry = new AtomicBoolean(true);
+    /**
+     * A method which returns a CompletableFuture once *all* of the existing pods for the given label are in the running
+     * state. This means that it will wait for terminated pods to be removed in the case of a scale down event before
+     * proceeding.
+     *
+     * @param namespace Namespace
+     * @param labelName Label name.
+     * @param labelValue Value of the Label.
+     * @param expectedPodCount Number of pods that need to be running.
+     * @param retries The number of retries to attempt.
+     * @return A future which completes once the number of running pods matches the given criteria.
+     */
+    public CompletableFuture<Void> waitUntilPodIsRunningRetries(String namespace, String labelName, String labelValue, int expectedPodCount, int retries) {
+        AtomicInteger retriesRemaining = new AtomicInteger(retries);
+        Supplier<Boolean> shouldRetry = () -> retriesRemaining.getAndDecrement() > 0;
 
-        return Futures.loop(shouldRetry::get,
+        return Futures.loop(shouldRetry,
                 () -> Futures.delayedFuture(Duration.ofSeconds(5), executor) // wait for 5 seconds before checking for status.
-                        .thenCompose(v -> getStatusOfPodWithLabel(namespace, labelName, labelValue)) // fetch status of pods with the given label.
-                        .thenApply(podStatuses -> podStatuses.stream()
-                                // check for pods where all containers are running.
-                                .filter(podStatus -> {
-                                    if (podStatus.getContainerStatuses() == null) {
-                                        return false;
-                                    } else {
-                                        return podStatus.getContainerStatuses()
-                                                .stream()
-                                                .allMatch(st -> st.getState().getRunning() != null);
-                                    }
-                                }).count()),
-                runCount -> { // Number of pods which are running
-                    log.info("Expected running pod count of {}:{}, actual running pod count of {}:{}.", labelValue, expectedPodCount, labelValue, runCount);
-                    if (runCount == expectedPodCount) {
-                        shouldRetry.set(false);
+                        .thenCompose(v -> getPodsWithLabel(namespace, labelName, labelValue)) // fetch status of pods with the given label.
+                        .thenApply(pods -> pods),
+                pods -> {
+                    long runCount = pods.getItems().stream().map(pod -> pod.getStatus()).filter(this::isPodRunning).count();
+                    long totalCount = pods.getItems().size();
+
+                    if (runCount == totalCount && runCount == expectedPodCount) {
```

Comment:
Would it be the same as saying `totalCount == expectedPodCount`?

Reply:
No, because the `totalCount` will not always be the same value as the `runCount`. If we are expecting 3 pods to be running, but only have two running pods and another pod that has been terminated (another one has yet to be spun up, and the terminated pod yet to be cleaned up) we will still have a `totalCount` of 3, an `expectedPodCount` of 3 and a `runCount` of 2.